File: web/src/shared/api/referral.ts
import { get, post, put } from './client';
import type { PagedData, PageReq } from '../types';

// 用户侧「我的邀请」概览
export interface MyReferralResp {
  invite_code: string;
  // 邀请链接前缀（后台可配）；空 = 用当前控制台域名
  link_base_url: string;
  enabled: boolean;
  invitee_count: number;
  total_rebate: number;
  total_reversed: number;
}

// 用户侧返利流水（被邀请人邮箱已脱敏）
export interface MyReferralCommissionResp {
  id: number;
  invitee_email: string;
  paid_amount: number;
  rate: number;
  amount: number;
  status: 'settled' | 'reversed';
  created_at: string;
  reversed_at?: string;
}

// 管理端返利流水（完整字段）
export interface ReferralCommissionResp {
  id: number;
  inviter_id: number;
  inviter_email: string;
  invitee_id: number;
  invitee_email: string;
  out_trade_no: string;
  kind: 'rebate' | 'first_bonus';
  paid_amount: number;
  rate: number;
  amount: number;
  status: 'settled' | 'reversed';
  created_at: string;
  reversed_at?: string;
}

// 推广官汇总行（对账报表）
export interface ReferralPromoterResp {
  user_id: number;
  email: string;
  username: string;
  referral_rate: number | null;
  invitee_count: number;
  total_rebate: number;
  total_reversed: number;
  first_bonus_total: number;
}

export const referralApi = {
  // 用户接口
  me: () => get<MyReferralResp>('/api/v1/referral/me'),
  myCommissions: (params: PageReq) =>
    get<PagedData<MyReferralCommissionResp>>('/api/v1/referral/commissions', params),

  // 管理员接口
  summary: () => get<ReferralPromoterResp[]>('/api/v1/admin/referral/summary'),
  commissions: (params: PageReq & { inviter_id?: number; invitee_id?: number; kind?: string; status?: string }) =>
    get<PagedData<ReferralCommissionResp>>('/api/v1/admin/referral/commissions', params),
  reverse: (id: number) =>
    post<ReferralCommissionResp>(`/api/v1/admin/referral/commissions/${id}/reverse`),
  // rate 传 null 清除用户级覆盖（回落全局默认）
  setUserRate: (userId: number, rate: number | null) =>
    put<void>(`/api/v1/admin/referral/users/${userId}/rate`, { rate }),
};

File: backend/internal/server/dto/referral.go
package dto

import "time"

// MyReferralResp 用户侧「我的邀请」概览。
type MyReferralResp struct {
	InviteCode string `json:"invite_code"`
	// LinkBaseURL 邀请链接前缀（后台可配）；空 = 前端用当前控制台域名拼接。
	LinkBaseURL string `json:"link_base_url"`
	Enabled     bool   `json:"enabled"`
	// ReferralRate 当前用户有效返利比例（0~1）：用户级覆盖 else 全局默认。
	ReferralRate  float64 `json:"referral_rate"`
	InviteeCount  int     `json:"invitee_count"`
	TotalRebate   float64 `json:"total_rebate"`
	TotalReversed float64 `json:"total_reversed"`
	// Tier 推广身份：user / official（驱动页面官方推广官徽章样式）。
	Tier string `json:"tier"`
	// DisplayName 官方推广官署名（仅 official 有意义）。
	DisplayName string `json:"display_name"`
}

// ReferralResolveResp 邀请码解析（公开端点，供注册页/落地页渲染访客侧认证条）。
type ReferralResolveResp struct {
	Exists      bool   `json:"exists"`
	Tier        string `json:"tier"`
	DisplayName string `json:"display_name"`
	BadgeText   string `json:"badge_text"`
}

// MyReferralCommissionResp 用户侧返利流水（被邀请人邮箱已脱敏，不含订单号）。
type MyReferralCommissionResp struct {
	ID           int        `json:"id"`
	InviteeEmail string     `json:"invitee_email"`
	PaidAmount   float64    `json:"paid_amount"`
	Rate         float64    `json:"rate"`
	Amount       float64    `json:"amount"`
	Status       string     `json:"status"`
	CreatedAt    time.Time  `json:"created_at"`
	ReversedAt   *time.Time `json:"reversed_at,omitempty"`
}

// ReferralCommissionResp 管理端返利流水（完整字段）。
type ReferralCommissionResp struct {
	ID           int        `json:"id"`
	InviterID    int        `json:"inviter_id"`
	InviterEmail string     `json:"inviter_email"`
	InviteeID    int        `json:"invitee_id"`
	InviteeEmail string     `json:"invitee_email"`
	OutTradeNo   string     `json:"out_trade_no"`
	Kind         string     `json:"kind"`
	PaidAmount   float64    `json:"paid_amount"`
	Rate         float64    `json:"rate"`
	Amount       float64    `json:"amount"`
	Status       string     `json:"status"`
	CreatedAt    time.Time  `json:"created_at"`
	ReversedAt   *time.Time `json:"reversed_at,omitempty"`
}

// ReferralPromoterResp 推广官汇总行（管理端对账报表）。
type ReferralPromoterResp struct {
	UserID          int      `json:"user_id"`
	Email           string   `json:"email"`
	Username        string   `json:"username"`
	ReferralRate    *float64 `json:"referral_rate"`
	InviteeCount    int      `json:"invitee_count"`
	TotalRebate     float64  `json:"total_rebate"`
	TotalReversed   float64  `json:"total_reversed"`
	FirstBonusTotal float64  `json:"first_bonus_total"`
	Tier            string   `json:"tier"`         // user / official
	DisplayName     string   `json:"display_name"` // 官方推广官署名
	InviteCode      string   `json:"invite_code"`  // 本人邀请码（官方为品牌 vanity 码）
}

// ReferralCommissionListReq 管理端流水筛选。
type ReferralCommissionListReq struct {
	PageReq
	InviterID int    `form:"inviter_id" binding:"omitempty,min=1"`
	InviteeID int    `form:"invitee_id" binding:"omitempty,min=1"`
	Kind      string `form:"kind" binding:"omitempty,oneof=rebate first_bonus"`
	Status    string `form:"status" binding:"omitempty,oneof=settled reversed"`
}

// SetReferralRateReq 设置用户级返利比例覆盖；rate 传 null 表示清除覆盖（回落全局默认）。
type SetReferralRateReq struct {
	Rate *float64 `json:"rate" binding:"omitempty,gte=0,lte=1"`
}

// SetPromoterReq 设置推广身份（官方/普通）。仅改展示样式，不动返佣比例。
// official=true 授予官方；invite_code 非空则覆盖设为品牌 vanity 码（4~16 位字母数字）；
// display_name 为官方推广官对访客展示的团队/署名。
type SetPromoterReq struct {
	Official    bool   `json:"official"`
	InviteCode  string `json:"invite_code" binding:"omitempty,max=16"`
	DisplayName string `json:"display_name" binding:"omitempty,max=64"`
}
